skills/types-check/SKILL.md

```diff
@@ -0,0 +1,279 @@
+
+```r
+# Entry point: let defaults work
+#' @export
+my_function <- function(x, name) {
+  check_string(x)        # Correct! Defaults capture user's context
+  check_name(name)       # Correct!
+
+  # ... function body
+}
+```
+
+### Check wrapper functions: DO pass `arg`/`call`
+
+When creating a wrapper or helper function that calls `check_*` functions on behalf of another function, you **must** propagate the caller context. Otherwise, errors will point to your wrapper function instead of the actual entry point.
+
+Without proper propagation, error messages show the wrong function and argument names:
+
+```r
+# WRONG: errors will point to check_positive's definition
+check_positive <- function(x) {
+  check_number_whole(x, min = 1)
+}
+
+my_function <- function(count) {
+  check_positive(count)
+}
+
+my_function(-5)
+#> Error in `check_positive()`:  # Wrong! Should say `my_function()`
+#> ! `x` must be a whole number larger than or equal to 1.  # Wrong! Should say `count`
+```
+
+With proper propagation, errors correctly identify the entry point and argument:
+
+```r
+# CORRECT: propagates context from the entry point
+check_positive <- function(x, arg = caller_arg(x), call = caller_env()) {
+  check_number_whole(x, min = 1, arg = arg, call = call)
+}
+
+my_function <- function(count) {
+  check_positive(count)
+}
+
+my_function(-5)
+#> Error in `my_function()`:  # Correct!
+#> ! `count` must be a whole number larger than or equal to 1.  # Correct!
+```
+
+## Other useful checkers
+
+Exported from other packages:
+
+- `rlang::arg_match()`: Validates enumerated choices. Partial matching is an error unlike `base::match.arg()`. Use when an argument must be one of a known set of strings.
+
+  ```r
+  # Validates and returns the matched value
+  my_plot <- function(color = c("red", "green", "blue")) {
+  color <- rlang::arg_match(color)
+  # ...
+  }
+
+  my_plot("redd")
+  #> Error in `my_plot()`:
+  #> ! `color` must be one of "red", "green", or "blue", not "redd".
+  #> ℹ Did you mean "red"?
+  ```
+
+- `rlang::check_required()`: Nice error message if required argument is not supplied.
+
+- `vctrs::obj_check_list()` checks that input is considered a list in the vctrs sense:
+  - A bare list with no class
+  - A list explicitly inheriting from `"list"`.
+
+- `vctrs::obj_check_vector()` checks that input is considered a vector in the vctrs sense:
+  - A base atomic type
+  - A list in the vctrs sense
+  - An object with a `vec_proxy()` method
+
+- `vctrs::vec_check_size(x, size)` tests if vector `x` has size `size`, and throws an informative error if it doesn't.
+
+- `vctrs::list_check_all_vectors()` and `vctrs::list_check_all_size()` check that inputs are lists containing only vectors, or only vectors of a given size.
```

Comment:
IMO the vctrs checks are sufficiently specialised that they should go elsewhere.